### ISSUE TITLE
Add User-Defined Functions

### ISSUE THREAD

**Issue by George-lewis** [George-lewis/rustcalc#4]
# Description

Add user-defined functions to the program. These functions can be defined by users at runtime and used in statements and expressions.

### Motivation

Allowing users to define their own functions makes the program much more flexible and allows for a lot more expressivity.

## Example

```
> def square a = a*a
[ TBD ]
> four = #square 2
[ $four = #square(2) ] => 4.000
> $four
[ $four ] => 4.000
```

```
> def pow a b = a ^ b
[ TBD ]
> #pow(3, 2)
[ #pow(3, 2) ] => 9.000
```

## Details

A new struct would be created:

```rust
struct Function {
  name: String,
  code: String,
  args: Vec<String>
}
```

`doeval`, `tokenize`, and `eval` would all have to accept a `&[UserFunction]` supplied by the caller. They would be tokenized by first recognizing the `#` prefix then their name. During evaluation a recursive call to `doeval` would be invoked where the code of the function is the input, and the variables are the same as the current context plus those defined on the function struct.

---

Details are subject to change at any time and are up for discussion.

**Comment by George-lewis** [comment]
There's a really dirty implementation of this in branch [functions](https://github.com/George-lewis/rustcalc/tree/functions) right now. I need to think about the structure a bit more and wait until the variables stuff is complete.

**Comment by George-lewis** [comment]
closed by #13
